src/gs-page.c

```diff
@@ -481,6 +515,45 @@ gs_page_install_app (GsPage *page,
 	}
 }
 
+void
+gs_page_copy_app (GsPage *page,
+		  GsApp *app,
+		  const gchar *dest,
+		  GsShellInteraction interaction,
+		  GCancellable *cancellable)
+{
+	GsPagePrivate *priv = gs_page_get_instance_private (page);
+	GsPageHelper *helper;
+
+	helper = g_slice_new0 (GsPageHelper);
+	helper->action = GS_PLUGIN_ACTION_COPY;
+	helper->app = g_object_ref (app);
+	helper->page = g_object_ref (page);
+	helper->cancellable = g_object_ref (cancellable);
+	helper->interaction = interaction;
+
+	if (gs_app_get_state (app) == AS_APP_STATE_INSTALLED) {
+		g_autoptr(GsPluginJob) plugin_job = NULL;
+
+		plugin_job = gs_plugin_job_newv (helper->action,
+						 "app", helper->app,
+						 "dest", dest,
+						 "failure-flags", GS_PLUGIN_FAILURE_FLAGS_USE_EVENTS |
+								  GS_PLUGIN_FAILURE_FLAGS_FATAL_ANY,
+						 NULL);
+
+		gs_plugin_loader_job_process_async (priv->plugin_loader,
+						    plugin_job,
+						    helper->cancellable,
+						    gs_page_app_copied_cb,
+						    helper);
+	} else {
+		// NOTE: it should be impossible to reach this state
```

Comment:
I think it should be unlikely this is reached, but the states here are not atomic, so things like this are usually reachable in g-s :sweat_smile: